### PR TITLE
Refactor Symop to be multiple classes

### PR DESCRIPTION
**Pull request by oerc0122**

### CODE REVIEW

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: horace_core/symop/@Symop/Symop.m [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+            end
+            obj.offset_ = val(:);
+        end
+
+        function obj = set.W(obj, val)
+            if  ~obj.is_3x3matrix(val) || abs(det(val)) ~= 1
+                error('HORACE:symop:invalid_argument', ...
+                      'Motion matrix W must be a 3x3 matrix with determinant 1, det: %d', det(val));
+            end
+            obj.W_ = reshape(val, [3 3]);
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply horace_core/symop/@Symop/Symop.m:84]: It just needs to have 9 elements in any shape as defined by `is_3x3matrix` (requires `isnumeric && numel == 9`)

File: horace_core/symop/@Symop/Symop.m [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
```
Comment (cmarooney-stfc) [inline horace_core/symop/@Symop/Symop.m:97]:
```suggestion
            obj.W_ = reshape(val, [3 3]); % as tested by is_3x3matrix
```

File: horace_core/sqw/@SQWDnDBase/private/cut_parse_inputs_.m [hunk 3]
```diff
@@ -336,20 +375,69 @@
...
-end
\ No newline at end of file
+end
+
+end
+
+function sym_out = check_sym_arg(sym)
+% Checks on symmetry description - check valid, and remove empty descriptions
+%
+%   >> sym_out = cut_sqw_check_sym_arg (sym)
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@SQWDnDBase/private/cut_parse_inputs_.m:401]: check name

File: horace_core/sqw/@SQWDnDBase/private/cut_parse_inputs_.m [hunk 4]
```diff
@@ -336,20 +375,69 @@
...
+    if ~isempty(sym{i}) && ~isa(sym{i}, 'Symop')
+        error('HORACE:cut_sqw_sym:invalid_argument', ...
+              'Symmetry descriptor must be an symop object or array of symop objects, or a cell of those');
+    end
+
+    keep(i) = ~isempty(sym{i}) && ...
+              ~all(arrayfun(@(x) isa(x, 'SymopIdentity'), sym{i}));
+end
+
+% Always return identity
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@SQWDnDBase/private/cut_parse_inputs_.m:440]:
```suggestion
% Always add identity in addition to other kept symops
```

File: horace_core/symop/@Symop/Symop.m [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+% --------------------------------------
+%   symop           - Create a symmetry operator object
+%   transform_vec   - Transform a 3xN list of vectors
+%   transform_pix   - Transform pixel coordinates into symmetry related coordinates
+%   transform_proj  - Transform projection axes description by the symmetry operation
+
+    properties (Dependent)
+        offset;
+        W;
+    end
```
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply horace_core/symop/@Symop/Symop.m:47]: did not read Dependent correctly, this is ok

File: horace_core/symop/@Symop/Symop.m [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+            if ~exist('offset', 'var')
+                offset = obj.offset;
+            end
+
+            if ~Symop.check_args({W, offset})
+                error('HORACE:symop:invalid_argument', ...
+                      ['Constructor arguments should be:\n', ...
+                       '- Motion:     symop(3x3matrix, [3vector])\n', ...
+                       'Received: %s'], disp2str(W));
```
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply horace_core/symop/@Symop/Symop.m:67]: fixed by JW

File: horace_core/symop/@Symop/Symop.m [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+            if  ~obj.is_3x3matrix(val) || abs(det(val)) ~= 1
+                error('HORACE:symop:invalid_argument', ...
+                      'Motion matrix W must be a 3x3 matrix with determinant 1, det: %d', det(val));
+            end
+            obj.W_ = reshape(val, [3 3]);
+        end
+
+        function W = get.W(obj)
+            W = obj.transform_vec(eye(3));
+        end
```
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply horace_core/symop/@Symop/Symop.m:91]: fixed by JW

File: horace_core/symop/@Symop/Symop.m [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+            end
+
+            for i = numel(obj):-1:1
+                R = obj(i).calculate_transform(eye(3));
+                vec = R * vec;
+            end
+        end
+
+        function disp(obj)
+            if isscalar(obj)
```
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply horace_core/symop/@Symop/Symop.m:128]: L 127 fixes

File: horace_core/symop/@Symop/Symop.m [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+        %       axis    Vector defining the rotation axis
+        %               (in reciprocal lattice units: (h,k,l))
+        %       angle   Angle of rotation in degrees
+        %       offset  [Optional] Vector defining a point in reciprocal lattice units
+        %               through which the rotation axis passes
+        %               Default: [0,0,0] i.e. the rotation axis goes throught the origin
+        %
+        %   Reflection:
+        %       >> obj = Symop.create (u, v)
+        %       >> obj = Symop.create (u, v, offset)
```
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply horace_core/symop/@Symop/Symop.m:296]: by argument types (jw)

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

**Pull request merged**